lib/hobbes/boot/eqord.hob

```diff
@@ -74,43 +74,51 @@ instance Ord datetime datetime where
   x >  y = (convert(x)::long) >  (convert(y)::long)
   x >= y = (convert(x)::long) >= (convert(y)::long)
 
-instance (Ord a a) => Ord a@f a@f where
-  x <  y = load(x) < load(y)
-  x <= y = load(x) <= load(y)
-  x >  y = load(x) >  load(y)
-  x >= y = load(x) >= load(y)
+instance (Ord a b) => Ord a@f b where
+  x <  y = load(x) <  y
+  x <= y = load(x) <= y
+  x >  y = load(x) >  y
+  x >= y = load(x) >= y
+
+instance (Ord a b) => Ord a b@f where
+  x <  y = x <  load(y)
+  x <= y = x <= load(y)
+  x >  y = x >  load(y)
+  x >= y = x >= load(y)
 
 // variant ordering
-instance (Ord a a) => Ord a (a+b) where
-  x <  y = case y of |0:yv=x <  yv,1:yv=false|
-  x <= y = case y of |0:yv=x <= yv,1:yv=false|
-  x >  y = case y of |0:yv=x >  yv,1:yv=false|
-  x >= y = case y of |0:yv=x >= yv,1:yv=false|
-
-instance (Ord a a) => Ord (a+b) a where
-  x <  y = y >  x
-  x <= y = y >= x
-  x >  y = y <  x
-  x >= y = y <= x
-
-instance (Ord b b) => Ord b (a+b) where
-  x <  y = case y of |0:yv=false,1:yv=x <  yv|
-  x <= y = case y of |0:yv=false,1:yv=x <= yv|
-  x >  y = case y of |0:yv=false,1:yv=x >  yv|
-  x >= y = case y of |0:yv=false,1:yv=x >= yv|
-
-instance (Ord b b) => Ord (a+b) b where
-  x <  y = y >  x
-  x <= y = y >= x
-  x >  y = y <  x
-  x >= y = y <= x
-
-instance (Ord a a, Ord b b) => Ord (a+b) (a+b) where
+instance (Ord a c, Ord b d) => Ord (a+b) (c+d) where
   x <  y = (match x y with | |0=ax| |0=ay| -> ax <  ay | |1=bx| |1=by| -> bx <  by | |0=_| _ -> true  | _ _ -> false)
```

Comment:
i think it's better to update `ay` to `cy` and `by` to `dy`.
what's the rationale to have `|0=_| _ -> true` ?

Reply:
Sure, might as well make the names line up with the types.

The reason for the `|0=_| _ -> true` case just has to do with how you define variants (here the special case of a two-way sum type) when the constructors don't line up.  So after we've established that the pair of variants being considered aren't both the 0 case, and they're not both the 1 case, then it must be that one is the 0 case and the other is the 1 case.  Here we have to define an ordering of constructor IDs (not values, since we don't have 'Ord a d' and 'Ord b c').  So having this case `|0=_| _ -> true` literally says that all of the 0 cases of the first variant are less than all of the 1 cases of the second, and the last bit `_ _ -> false` says that all of the 1 cases of the first variant are not less than all of the 0 cases of the second.

An important case to keep in mind for this is sorting, so we can consider just one sum type.  When we define ordering this way, and we sort an array of a sum type, then the result will be all of the 0 cases (in order of the payload type for the 0 case) followed by all of the 1 cases (in order of the payload type for the 1 case).

```
> sort([ just(42), nothing, just(7), nothing ])
==>
[nothing, nothing, just(7), just(42)]
```

Reply:
thanks for the explanation!